File: k9mail/src/main/java/com/fsck/k9/notification/AddNotificationResult.java
package com.fsck.k9.notification;


class AddNotificationResult {
    private final NotificationHolder notificationHolder;
    private final boolean cancelNotificationBeforeReuse;


    private AddNotificationResult(NotificationHolder notificationHolder,
            boolean cancelNotificationBeforeReuse) {
        this.notificationHolder = notificationHolder;
        this.cancelNotificationBeforeReuse = cancelNotificationBeforeReuse;
    }

    public static AddNotificationResult newNotification(NotificationHolder notificationHolder) {
        return new AddNotificationResult(notificationHolder, false);
    }

    public static AddNotificationResult replaceNotification(NotificationHolder notificationHolder) {
        return new AddNotificationResult(notificationHolder, true);
    }

    public boolean shouldCancelNotification() {
        return cancelNotificationBeforeReuse;
    }

    public int getNotificationId() {
        if (!shouldCancelNotification()) {
            throw new IllegalStateException("getNotificationId() can only be called when " +
                    "shouldCancelNotification() returns true");
        }

        return notificationHolder.notificationId;
    }

    public NotificationHolder getNotificationHolder() {
        return notificationHolder;
    }
}

File: app/core/src/main/java/com/fsck/k9/notification/AddNotificationResult.kt
package com.fsck.k9.notification

internal class AddNotificationResult private constructor(
    val notificationData: NotificationData,
    val notificationHolder: NotificationHolder,
    val shouldCancelNotification: Boolean
) {
    val cancelNotificationId: Int
        get() {
            check(shouldCancelNotification) { "shouldCancelNotification == false" }
            return notificationHolder.notificationId
        }

    companion object {
        fun newNotification(
            notificationData: NotificationData,
            notificationHolder: NotificationHolder
        ): AddNotificationResult {
            return AddNotificationResult(notificationData, notificationHolder, shouldCancelNotification = false)
        }

        fun replaceNotification(
            notificationData: NotificationData,
            notificationHolder: NotificationHolder
        ): AddNotificationResult {
            return AddNotificationResult(notificationData, notificationHolder, shouldCancelNotification = true)
        }
    }
}
